dandi/validate.py

```diff
@@ -44,12 +90,101 @@ def validate_bids(
         report_path = report_path.format(
             log_dir=log_dir,
         )
+
     validation_result = validate_bids_(
         paths,
         schema_version=schema_version,
         report_path=report_path,
     )
-    return dict(validation_result)
+    our_validation_result = []
+    origin = ValidationOrigin(
+        name="bidsschematools",
+        version=bidsschematools.__version__,
+    )
+
+    for path in validation_result["path_tracking"]:
+        # Hard-coding exclusion here pending feature + release in:
+        # https://github.com/bids-standard/bids-specification/issues/1272
+        if path.endswith((".ERRORS", ".ERRORS.json")):
+            continue
+        dataset_path = _get_set_path(path, "dataset_description.json")
+        dandiset_path = _get_set_path(path, "dandiset.yaml")
+        our_validation_result.append(
+            ValidationResult(
+                origin=origin,
+                severity=Severity.ERROR,
+                id="BIDS.NON_BIDS_PATH_PLACEHOLDER",
+                scope=Scope.FILE,
+                path=Path(path),
+                message="File does not match any pattern known to BIDS.",
+                dataset_path=dataset_path,
+                dandiset_path=dandiset_path,
+            )
+        )
+
+    for pattern in validation_result["schema_tracking"]:
+        # Future proofing for standard-compliant name.
+        if pattern.get("mandatory") or pattern.get("required"):
+            # We don't have a path for this so we'll need some external logic to make sure
+            # that the dataset path is populated.
+            # dataset_path = _get_dataset_path(path, paths)
+            our_validation_result.append(
+                ValidationResult(
+                    origin=origin,
+                    severity=Severity.ERROR,
+                    id="BIDS.MANDATORY_FILE_MISSING_PLACEHOLDER",
+                    scope=Scope.DATASET,
+                    path_regex=pattern["regex"],
+                    message="BIDS-required file is not present.",
+                )
+            )
+    for meta in validation_result["match_listing"]:
+        file_path = meta.pop("path")
+        meta = {BIDS_TO_DANDI[k]: v for k, v in meta.items() if k in BIDS_TO_DANDI}
+        dataset_path = _get_set_path(file_path, "dataset_description.json")
+        dandiset_path = _get_set_path(file_path, "dandiset.yaml")
+        # Top level files do not have any other metadata other than path,
+        # which we pop and put in the object...
+        if not meta:
+            meta = {
+                "subject": None,
+            }
```

Comment:
is `subject` mandatory? why?  why it cannot stay an empty `{}`?

Reply:
I'm pretty sure this used to fail somehow, but I can't reproduce the failure. Maybe something that got sorted out by the `assert`s. In any case, dealt with: b76b8bbbd94dfa3ebfc969ed08482369bf55d28b

Reply:
Ok, no, it is required, it's explained in the above comment. Some files only have path in `meta` and that gets popped since we store the path separately, meaning that `assert` evaluates it as false making it look like an invalid entry.

```
2022-09-26T21:25:58.7861309Z =================================== FAILURES ===================================
2022-09-26T21:25:58.7861775Z ___________________________ test_upload_bids_invalid ___________________________
2022-09-26T21:25:58.7862256Z dandi/tests/test_upload.py:188: in test_upload_bids_invalid
2022-09-26T21:25:58.7862710Z     bids_dandiset_invalid.upload(existing="force")
2022-09-26T21:25:58.7863110Z dandi/tests/fixtures.py:429: in upload
2022-09-26T21:25:58.7863499Z     **{**self.upload_kwargs, **kwargs},
2022-09-26T21:25:58.7863869Z dandi/upload.py:301: in upload
2022-09-26T21:25:58.7864249Z     for v in process_path(dfile):
2022-09-26T21:25:58.7864611Z dandi/upload.py:144: in process_path
2022-09-26T21:25:58.7865035Z     validation_errors = dfile.get_validation_errors()
2022-09-26T21:25:58.7865471Z dandi/files/bids.py:233: in get_validation_errors
2022-09-26T21:25:58.7865889Z     ) + BIDSAsset.get_validation_errors(self)
2022-09-26T21:25:58.7866304Z dandi/files/bids.py:158: in get_validation_errors
2022-09-26T21:25:58.7866738Z     return self.bids_dataset_description.get_asset_errors(self)
2022-09-26T21:25:58.7867175Z dandi/files/bids.py:89: in get_asset_errors
2022-09-26T21:25:58.7867537Z     self._validate()
2022-09-26T21:25:58.7867890Z dandi/files/bids.py:82: in _validate
2022-09-26T21:25:58.7868255Z     assert result.metadata
2022-09-26T21:25:58.7868580Z E   AssertionError
```

Reply:
well, without deeper analysis I cannot really judge on why we have that `assert` i.e. why we have assumption that metadata should have something!  may be it was supposed to be `assert result.metadata is not None` or smth like that?

Reply:
you're right aeb6d23ebbf1580e1c6ab8435431b8c92dfa932e